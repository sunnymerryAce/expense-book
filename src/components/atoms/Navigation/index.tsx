import React, { useEffect } from "react";
// ナビゲーションのみ例外的にatomからreduxへの接続を許容する
import { useSelector, useDispatch } from "react-redux";
import { push, RouterState } from "connected-react-router";
import BottomNavigation from "@material-ui/core/BottomNavigation";
import BottomNavigationAction from "@material-ui/core/BottomNavigationAction";
import BarChartIcon from "@material-ui/icons/BarChart";
import ListIcon from "@material-ui/icons/List";
import SettingsIcon from "@material-ui/icons/Settings";
import { RootState } from "@/state/store";
import useStyles from "./style";

export interface NavigationProps {
  pathname: string;
  route: (index: number) => void;
}

export const routes = [
  {
    path: "/",
    index: 0,
  },
  {
    path: "/list",
    index: 1,
  },
  {
    path: "/budget",
    index: 2,
  },
  {
    path: "/category",
    index: 3,
  },
];

export const Navigation: React.FC<NavigationProps> = (props) => {
  const bottomNavigationClasses = useStyles("BottomNavigation");
  const bottomNavigationActionClasses = useStyles("BottomNavigationAction");
  // デフォルトでは何もアクティブにしないよう-1を設定
  const [index, setIndex] = React.useState(-1);
  // 初回レンダリング時、パスからカレントのインデックスを設定する
  useEffect(() => {
    const current = routes.find((item) => {
      return item.path === props.pathname;
    });
    setIndex(current ? current.index : 0);
  }, []);

  return (
    <BottomNavigation
      value={index}
      onChange={(event, index): void => {
        setIndex(index);
        props.route(index);
      }}
      classes={bottomNavigationClasses}
      showLabels
    >
      <BottomNavigationAction
        label="グラフ"
        icon={<BarChartIcon />}
        classes={bottomNavigationActionClasses}
      />
      <BottomNavigationAction
        label="一覧"
        icon={<ListIcon />}
        classes={bottomNavigationActionClasses}
      />
      <BottomNavigationAction
        label="予算"
        icon={<SettingsIcon />}
        classes={bottomNavigationActionClasses}
      />
      {/* <BottomNavigationAction
        label="カテゴリ"
        icon={<SettingsIcon />}
        classes={bottomNavigationActionClasses}
      /> */}
    </BottomNavigation>
  );
};

export const ConnectedNavigation: React.FC = () => {
  const router = useSelector<RootState, RouterState>((state) => state.router);
  const dispatch = useDispatch();
  const _props: NavigationProps = {
    pathname: router.location.pathname,
    route: (index: number): void => {
      const to = routes.find((item) => {
        return item.index === index;
      });
      if (to) dispatch(push(to.path));
    },
  };
  return <Navigation {..._props} />;
};

export default ConnectedNavigation;
