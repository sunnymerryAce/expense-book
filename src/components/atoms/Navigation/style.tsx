import { makeStyles, createStyles, Theme } from "@material-ui/core/styles";

const useStyles = (module: string): Record<string, any> => {
  switch (module) {
    case "BottomNavigation":
      return makeStyles((theme: Theme) =>
        createStyles({
          root: {
            width: "100vw",
            background: theme.palette.primary.dark,
          },
        })
      )();
    case "BottomNavigationAction":
      return makeStyles((theme: Theme) => {
        return createStyles({
          selected: {
            // importantしないと優先度で負ける
            color: "#fff !important",
          },
          iconOnly: {
            color: theme.palette.primary.light,
          },
        });
      })();
    default:
      return makeStyles({})();
  }
};

export default useStyles;
