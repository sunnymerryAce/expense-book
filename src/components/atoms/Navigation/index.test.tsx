import React from "react";
import Navigation from "./";
import { shallow } from "enzyme";

describe("<Navigation />", () => {
  let container: any = null;
  beforeEach(() => {
    container = shallow(<Navigation />);
  });

  afterEach(() => {
    container.unmount();
    container = null;
  });

  test("should match the snapshot", () => {
    expect(container.html()).toMatchSnapshot();
  });
});
