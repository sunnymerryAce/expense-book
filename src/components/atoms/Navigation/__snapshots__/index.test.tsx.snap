// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navigation /> should match the snapshot 1`] = `"<div class=\\"MuiBottomNavigation-root makeStyles-root-1\\"><button class=\\"MuiButtonBase-root MuiBottomNavigationAction-root Mui-selected makeStyles-selected-2\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"MuiBottomNavigationAction-wrapper\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M5 9.2h3V19H5zM10.6 5h2.8v14h-2.8zm5.6 8H19v6h-2.8z\\"></path></svg><span class=\\"MuiBottomNavigationAction-label Mui-selected makeStyles-selected-2\\">Chart</span></span></button><button class=\\"MuiButtonBase-root MuiBottomNavigationAction-root\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"MuiBottomNavigationAction-wrapper\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M3 13h2v-2H3v2zm0 4h2v-2H3v2zm0-8h2V7H3v2zm4 4h14v-2H7v2zm0 4h14v-2H7v2zM7 7v2h14V7H7z\\"></path></svg><span class=\\"MuiBottomNavigationAction-label\\">List</span></span></button><button class=\\"MuiButtonBase-root MuiBottomNavigationAction-root\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"MuiBottomNavigationAction-wrapper\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path transform=\\"scale(1.2, 1.2)\\" d=\\"M15.95 10.78c.03-.25.05-.51.05-.78s-.02-.53-.06-.78l1.69-1.32c.15-.12.19-.34.1-.51l-1.6-2.77c-.1-.18-.31-.24-.49-.18l-1.99.8c-.42-.32-.86-.58-1.35-.78L12 2.34c-.03-.2-.2-.34-.4-.34H8.4c-.2 0-.36.14-.39.34l-.3 2.12c-.49.2-.94.47-1.35.78l-1.99-.8c-.18-.07-.39 0-.49.18l-1.6 2.77c-.1.18-.06.39.1.51l1.69 1.32c-.04.25-.07.52-.07.78s.02.53.06.78L2.37 12.1c-.15.12-.19.34-.1.51l1.6 2.77c.1.18.31.24.49.18l1.99-.8c.42.32.86.58 1.35.78l.3 2.12c.04.2.2.34.4.34h3.2c.2 0 .37-.14.39-.34l.3-2.12c.49-.2.94-.47 1.35-.78l1.99.8c.18.07.39 0 .49-.18l1.6-2.77c.1-.18.06-.39-.1-.51l-1.67-1.32zM10 13c-1.65 0-3-1.35-3-3s1.35-3 3-3 3 1.35 3 3-1.35 3-3 3z\\"></path></svg><span class=\\"MuiBottomNavigationAction-label\\">Budget</span></span></button><button class=\\"MuiButtonBase-root MuiBottomNavigationAction-root\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"MuiBottomNavigationAction-wrapper\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path transform=\\"scale(1.2, 1.2)\\" d=\\"M15.95 10.78c.03-.25.05-.51.05-.78s-.02-.53-.06-.78l1.69-1.32c.15-.12.19-.34.1-.51l-1.6-2.77c-.1-.18-.31-.24-.49-.18l-1.99.8c-.42-.32-.86-.58-1.35-.78L12 2.34c-.03-.2-.2-.34-.4-.34H8.4c-.2 0-.36.14-.39.34l-.3 2.12c-.49.2-.94.47-1.35.78l-1.99-.8c-.18-.07-.39 0-.49.18l-1.6 2.77c-.1.18-.06.39.1.51l1.69 1.32c-.04.25-.07.52-.07.78s.02.53.06.78L2.37 12.1c-.15.12-.19.34-.1.51l1.6 2.77c.1.18.31.24.49.18l1.99-.8c.42.32.86.58 1.35.78l.3 2.12c.04.2.2.34.4.34h3.2c.2 0 .37-.14.39-.34l.3-2.12c.49-.2.94-.47 1.35-.78l1.99.8c.18.07.39 0 .49-.18l1.6-2.77c.1-.18.06-.39-.1-.51l-1.67-1.32zM10 13c-1.65 0-3-1.35-3-3s1.35-3 3-3 3 1.35 3 3-1.35 3-3 3z\\"></path></svg><span class=\\"MuiBottomNavigationAction-label\\">Category</span></span></button></div>"`;
